dao/src/lib.rs

```diff
@@ -0,0 +1,692 @@
+        if !self.whitelist.contains(applicant) && !self.members.contains_key(applicant) {
+            panic!("Applicant must be either in whitelist or be a DAO member");
+        }
+
+        // transfer applicant tokens to DAO contract
+        transfer_tokens(
+            &self.approved_token_program_id,
+            applicant,
+            &exec::program_id(),
+            token_tribute,
+        )
+        .await;
+
+        let mut starting_period = exec::block_timestamp();
+        // compute startingPeriod for proposal
+        // there should be a minimum time interval between proposals (period_duration) so that members have time to ragequit
+        if self.proposal_id > 0 {
+            let previous_starting_period = self
+                .proposals
+                .get(&(&self.proposal_id - 1))
+                .unwrap()
+                .starting_period;
+            if starting_period < previous_starting_period + self.period_duration {
+                starting_period = previous_starting_period + self.period_duration;
+            }
+        }
+        let proposal = Proposal {
+            proposer: msg::source(),
+            applicant: *applicant,
+            shares_requested,
+            quorum,
+            is_membership_proposal: true,
+            token_tribute,
+            details,
+            starting_period,
+            ..Proposal::default()
+        };
+        self.proposals.insert(self.proposal_id, proposal);
+        msg::reply(
+            DaoEvent::SubmitMembershipProposal {
+                proposer: msg::source(),
+                applicant: *applicant,
+                proposal_id: self.proposal_id,
+                token_tribute,
+            },
+            0,
+        );
+        self.proposal_id = self.proposal_id.saturating_add(1);
+    }
+
+    // The proposal of funding
+    // Requirements:
+    // * The proposal can be submitted only by the existing members or their delegate addresses
+    // * The receiver ID can't be the zero
+    // * The DAO must have enough funds to finance the proposal
+    // Arguments:
+    // * `receiver`: an actor that will be funded
+    // * `amount`: the number of ERC20 tokens that will be sent to the receiver
+    // * `quorum`: a certain threshold of YES votes in order for the proposal to pass
+    // * `details`: the proposal description
+    async fn submit_funding_proposal(
+        &mut self,
+        applicant: &ActorId,
+        amount: u128,
+        quorum: u128,
+        details: String,
+    ) {
+        // check that `msg::source()` is either a DAO member or a delegate key
+        match self.member_by_delegate_key.get(&msg::source()) {
+            Some(member) => {
+                if !self.is_member(member) {
+                    panic!("account is not a DAO member");
+                }
+            }
+            None => {
+                panic!("account is not a delegate");
+            }
```

Comment:
No need for a block.
```suggestion
            None => panic!("account is not a delegate"),
```